File: String/ValidParentheses.java
import java.util.*;

class Solution {
    public boolean isValid(String s) {
        ArrayList<Character> stack = new ArrayList<>();
        char[] arr = s.toCharArray();
        for (char c : arr) {
            if (c == '(' || c == '[' || c == '{') {
              stack.add(c);
            } else {
              if (stack.isEmpty()) return false;
              char top = stack.get(stack.size() - 1);
              if ((top == '(' && c == ')') || (top == '[' && c == ']') || (top == '{' && c == '}')) {
                  stack.remove(stack.size() - 1);
              } else {
                  return false;
              }
            }
        }

        return stack.size() == 0;
    }
}

File: Python/String/ValidParentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        bracketMap = {'(' : ')', '[' : ']', '{' : '}'}
        for substr in s:
            if substr == '(' or substr == '[' or substr == '{': 
                stack.append(substr)
            else:
                if len(stack) == 0:
                    return False
                openStr = stack.pop()
                if bracketMap[openStr] != substr:
                    return False
        return len(stack) == 0
